File: src/pages/NavigationComponent.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from '@pages/base/BasePage';

/** Shared marketing-site navigation component migrated from Selenium's NavigationComponent. */
export class NavigationComponent extends BasePage {
  readonly carMenu: Locator;
  readonly bikeMenu: Locator;
  readonly healthMenu: Locator;
  readonly lifeMenu: Locator;

  constructor(page: Page) {
    super(page);
    this.carMenu = page.locator('a[href$="/car-insurance/"], a:has-text("Car Insurance")').first();
    this.bikeMenu = page.locator('a[href$="/bike-insurance/"], a:has-text("Bike Insurance")').first();
    this.healthMenu = page.locator('a[href$="/health-insurance/"], a:has-text("Health Insurance")').first();
    this.lifeMenu = page.locator('a[href$="/life-insurance/"], a:has-text("Life Insurance")').first();
  }

  async hoverOverCarMenu() { await this.carMenu.hover(); }
  async hoverOverBikeMenu() { await this.bikeMenu.hover(); }
  async hoverOverHealthMenu() { await this.healthMenu.hover(); }
  async hoverOverLifeMenu() { await this.lifeMenu.hover(); }

  private async clickMenuItem(menu: Locator, text: string) {
    await menu.hover();
    await this.page.getByRole('link', { name: text, exact: true }).click();
  }

  async navigateToCarInsuranceCompanies() { await this.clickMenuItem(this.carMenu, 'Car Insurance Companies'); }
  async navigateToCompareCarInsurance() { await this.clickMenuItem(this.carMenu, 'Compare Car Insurance'); }
  async navigateToBikePremiumCalculator() { await this.clickMenuItem(this.bikeMenu, 'Bike Insurance Premium Calculator'); }
  async navigateToHealthPremiumCalculator() { await this.clickMenuItem(this.healthMenu, 'Premium Calculator'); }
  async navigateToTermInsurance() { await this.clickMenuItem(this.lifeMenu, 'Term Insurance'); }

  async getAllTopLevelNavHrefs(): Promise<string[]> {
    const hrefs = await this.page.locator('nav a, header a').evaluateAll(links =>
      links.map(link => link.getAttribute('href')).filter((href): href is string => !!href && href.trim() !== '')
    );
    return [...new Set(hrefs)];
  }
}

File: src/main/java/com/turtlemint/pages/NavigationComponent.java
package com.turtlemint.pages;

import com.turtlemint.utils.WaitUtils;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;

import java.util.List;

/**
 * Represents the top mega-menu shared across all marketing pages:
 * Car / Bike / Health / Life, each with a dropdown of sub-links
 * (companies, product types, blogs, etc).
 *
 * Modeled as a component (not a full page) since it's reused across
 * every page object -- inject it into any page that needs nav actions
 * instead of duplicating these locators.
 */
public class NavigationComponent {

    private final WebDriver driver;
    private final WaitUtils wait;
    private final Actions actions;

    private final By carMenu = By.cssSelector("a[href$='/car-insurance/']");
    private final By bikeMenu = By.cssSelector("a[href$='/bike-insurance/']");
    private final By healthMenu = By.cssSelector("a[href$='/health-insurance/']");
    private final By lifeMenu = By.cssSelector("a[href$='/life-insurance/']");

    private final By carInsuranceCompaniesLink = By.linkText("Car Insurance Companies");
    private final By compareCarInsuranceLink = By.linkText("Compare Car Insurance");
    private final By bikePremiumCalculatorLink = By.linkText("Bike Insurance Premium Calculator");
    private final By healthPremiumCalculatorLink = By.linkText("Premium Calculator");
    private final By termInsuranceLink = By.linkText("Term Insurance");

    public NavigationComponent(WebDriver driver) {
        this.driver = driver;
        this.wait = new WaitUtils(driver);
        this.actions = new Actions(driver);
    }

    public void hoverOverCarMenu() {
        WebElement el = wait.waitForVisible(carMenu);
        actions.moveToElement(el).perform();
    }

    public void hoverOverBikeMenu() {
        WebElement el = wait.waitForVisible(bikeMenu);
        actions.moveToElement(el).perform();
    }

    public void hoverOverHealthMenu() {
        WebElement el = wait.waitForVisible(healthMenu);
        actions.moveToElement(el).perform();
    }

    public void navigateToCarInsuranceCompanies() {
        hoverOverCarMenu();
        wait.waitForClickable(carInsuranceCompaniesLink).click();
    }

    public void navigateToCompareCarInsurance() {
        hoverOverCarMenu();
        wait.waitForClickable(compareCarInsuranceLink).click();
    }

    public void navigateToBikePremiumCalculator() {
        hoverOverBikeMenu();
        wait.waitForClickable(bikePremiumCalculatorLink).click();
    }

    public void navigateToHealthPremiumCalculator() {
        hoverOverHealthMenu();
        wait.waitForClickable(healthPremiumCalculatorLink).click();
    }

    public void navigateToTermInsurance() {
        WebElement el = wait.waitForVisible(lifeMenu);
        actions.moveToElement(el).perform();
        wait.waitForClickable(termInsuranceLink).click();
    }

    /** Returns every top-level nav link's href -- useful for a bulk link-validity smoke test. */
    public List<String> getAllTopLevelNavHrefs() {
        return driver.findElements(By.cssSelector("nav a, header a"))
                .stream()
                .map(e -> e.getAttribute("href"))
                .filter(href -> href != null && !href.isBlank())
                .distinct()
                .toList();
    }
}
